### ISSUE TITLE
ContentChildren query lags in change detection when wrapped by template

### ISSUE THREAD

**Issue by waterplea** [angular/angular#38976]
# 🐞 bug report

### Affected Package

 The issue is caused by package @angular/core probably

### Is this a regression?

 No, the issue was present at least since version 7

### Description

When your component has `ng-template` in its content for later instantiation and inside that template there's a list of items — you can query those children with `ContentChildren` once you instantiate the template. However change detection in them lags a cycle.

## 🔬 Minimal Reproduction

 https://stackblitz.com/edit/angular-content-children-query-bug

## 🌍  Your Environment

**Angular Version:**
<pre><code>10</code></pre>

**Anything else relevant?**
In the reproduction Stackblitz, type a short string into input. 
This string is then broken into chars and spread into a list of DIVs with a 2 seconds delay.
You can even see that "content checked" is printed to console when the list updates.
However the `QueryList` of children is not refreshed.
Also note that I had to wrap `setInterval` into `runOutsideAngular`. 
Otherwise it would create a tick that refreshes the query, even though both components are `OnPush`.

**Comment by waterplea** [comment]
Cannot think of any workarounds :( Maybe there's a way to manually kick it to refresh? At least with some private API, it's blocking me at the moment :(

**Comment by Achilles1515** [comment]
If you have access to the component the template is defined in (`AppComponent` in this case) and set the change detection to `Default`, instead of `OnPush`, then it appears to work as expected. So I think some part of this issue has to do with the template defined view being `OnPush`/`CheckOnce`.

> Also note that I had to wrap `setInterval` into `runOutsideAngular`.
> Otherwise it would create a tick that refreshes the query, even though both components are `OnPush`.
I don't think this is the case.

If I comment out the `runOutsideAngular` like this:

```ts
export class HelloComponent {
  //...

  constructor(ngZone: NgZone) {
    // ngZone.runOutsideAngular(() => {
      setInterval(() => {
        console.log(this.children.length);
      }, 1000);
    // });
  }
  //...
}
```

...and type something in quickly, two seconds later after the observable emits I am still seeing `0` being reported as the content children length (0 until I type something else in).

Something is definitely up here.

**Comment by waterplea** [comment]
You are right about the zone. For some reason it did trigger it in my app where I originally stumbled upon it. Switching to `Default` is not an option for me. And since you see in console content is being checked, change detection should notice this in `OnPush` as well.

**Comment by waterplea** [comment]
Even `changeDetectorRef.detectChanges()` on the wrapping view doesn't seem to help. I suppose because content is technically a part of parent view.

**Comment by Achilles1515** [comment]
Adding:

```ts
ngDoCheck() {
    this._cref.detectChanges();
}
```

to `HelloComponent` works. Very ugly though.
If you wanted to drop into private APIs you could add a check to only detect changes if the component view is marked dirty, which should work for this specific example because the async pipe is marking the view dirty.

```ts
ngDoCheck() {
  const isDirty = (this as any).__ngContext__[2] & 0b000001000000;
  if (isDirty) {
    this._cref.detectChanges();
  }    
}
```

But if we're talking about private APIs...then there may be a more performant solution.
Queries are in dire need of work.

[Here is a modified StackBlitz](https://stackblitz.com/edit/angular-content-children-query-bug-xyuad3?file=src%2Fapp%2Fhello.component.ts) that adds the same query for `"test"` elements to the parent view (but, `ViewChildren` now), which more explicitly illustrates the problem.

The `ViewChildren` and `ContentChildren` queries are matching the same elements, but type something in the input and two seconds later they will have different `queryList.length` values as reported in the console.

**Comment by jelbourn** [comment]
@atscott is this potentially related transplated view change detection?

**Comment by atscott** [comment]
@jelbourn Yea, this is related to transplanted views. Ironically, there was a short period of time that this actually worked (change the version to `9.1.0` in the demo), but it worked in a way that broke other scenario (Transplanted views would get refreshed twice if both insertion and declaration are dirty. This could be an error if the changes in the insertion component result in data not being available to the transplanted view because it is slated to be removed.). 

Issue summary: 
* The `ng-template` is what we call a "transplanted view" - it's defined in one component and inserted into another
* We refresh transplanted views at their insertion points only. In this example, the content of the `ng-template` is refreshed as part of CD for `HelloComponent`.
* Content queries are refreshed as part of the _declaration_ component CD. Here, that means the content query gets refreshed when `AppComponent` gets refreshed. This happens before child components are refreshed. That means we're refreshing the content query before the new items are present.

My guess is that the implementation of content queries didn't consider that the ng-content might have transplanted views. I would also guess that fixing this would end up being a breaking change. The content query results should be available before executing the template of the insertion component or we could otherwise run into `ExpressionChangedAfterItWasChecked` issues.

**Comment by alxhub** [comment]
I consider this fixed by reactive content queries, which are now the recommended way to do content queries in Angular. I modernized the reproduction: https://stackblitz.com/edit/cq-signals-bug-fix?file=src%2Fmain.ts and change detection can even be used to show the count directly in the template (without `ExpressionChanged`).

Note that the original `ngDoCheck` logging _does_ still lag behind, as it runs before the `@for` loop creates the new content item. However, this is not a problem since signals ensure that as long as the query result is used reactively, the update will eventually be resolved.
